### PR TITLE
osd: improve hobject_t::to_str() performance

### PR DESCRIPTION
**Pull request by ronen-fr**
hobject_t::to_str() is used multiple times, mostly in scrub & snap-mapping code.
The modified version proposed here improves its performance , reducing
execution time (as measured) to about 70-75% of the original.

Also included:
- shaving ~65% of hobject fmtlib formatter execution time;
- additional minor improvements to the formatting of related
structures.

### CODE REVIEW

**Comment by ronen-fr** [comment]
(Crimson does not compile)

File: src/common/hobject_fmt.h [hunk 1]
```diff
@@ -39,14 +47,13 @@ template <> struct fmt::formatter<hobject_t> {
...
+    fmt::format_to(
+	ctx.out(), FMT_COMPILE("{}:{:08x}:"), static_cast<uint64_t>(ho.pool),
+	ho.get_bitwise_key_u32());
+    append_sanitized(ctx, ho.nspace, ':');
+    append_sanitized(ctx, ho.get_key(), ':');
+    append_sanitized(ctx, ho.oid.name);
+    return fmt::format_to(ctx.out(), FMT_COMPILE(":{}"), ho.snap);
   }
 };
+}  // namespace fmt
```
Comment (athanatos) [inline src/common/hobject_fmt.h]: Are there unit tests to guarantee that this doesn't change the resulting string?

**Review by athanatos — DISMISSED** [review]

File: src/common/hobject_fmt.h (hunk 1 above)
Comment (ronen-fr) [reply src/common/hobject_fmt.h]: There already is some external test code that I used during development (linked in commit messages).
I will add a unit test executable to this PR, as well.

File: src/common/hobject.cc [hunk 2]
```diff
@@ -14,9 +16,10 @@ using std::string;
 using ceph::bufferlist;
 using ceph::Formatter;
 
-static void append_escaped(const string &in, string *out)
+namespace {
+static void escape_special_chars(const string& in, string* out)
 {
-  for (string::const_iterator i = in.begin(); i != in.end(); ++i) {
```
Comment (Matan-B) [inline src/common/hobject.cc:19]: (const auto i : in)

File: src/common/hobject.cc [hunk 3]
```diff
@@ -14,9 +16,10 @@ using std::string;
 using ceph::bufferlist;
 using ceph::Formatter;
 
-static void append_escaped(const string &in, string *out)
+namespace {
+static void escape_special_chars(const string& in, string* out)
```
Comment (Matan-B) [inline src/common/hobject.cc:20]: IIUC static is reduandent in an unnamed namespace

File: src/common/hobject.cc [hunk 4]
```diff
@@ -80,33 +84,27 @@ set<string> hobject_t::get_prefixes(
...
+      return fmt::format(FMT_COMPILE("{:016X}.{:08X}.head."), poolid, revhash);
+    } else if (snap == CEPH_SNAPDIR) {
+      return fmt::format(
+	  FMT_COMPILE("{:016X}.{:08X}.snapdir."), poolid, revhash);
+    } else {
+      return fmt::format(
+	  FMT_COMPILE("{:016X}.{:08X}.{:X}."), poolid, revhash,
+	  (unsigned long long)snap);
+    }
+  }();
```
Comment (Matan-B) [inline src/common/hobject.cc:101]: nit: This works, although I think that we should avoid the complexity here and use a switch case instead.

File: src/test/objectstore/test_transaction.cc [hunk 5]
```diff
@@ -153,7 +152,6 @@ TEST(Transaction, GetNumBytes)
 
   coll_t acid;
   object_t aobj("another_test_name");
-  snapid_t asnap(0);
```
Comment (Matan-B) [inline src/test/objectstore/test_transaction.cc:156]: :+1:

**Review by Matan-B — COMMENTED** [review]
LGTM, few nits

File: src/common/hobject.cc (hunk 2 above)
Comment (Matan-B) [reply src/common/hobject.cc:19]: The ranged for loop can be used in the updated line

File: src/common/hobject.cc (hunk 4 above)
Comment (ronen-fr) [reply src/common/hobject.cc:101]: Agree. Modified to assigning into an empty out (I don't think a switch case adds clarity here, and the code produced should be the same).

File: src/test/common/test_hobject.cc [hunk 6]
```diff
@@ -9,3 +16,303 @@ TEST(HObject, cmp)
...
+      out->push_back('%');
+      out->push_back('u');
+    } else {
+      out->push_back(*i);
+    }
+  }
+}
+
+// original Ceph code as it was in version Squid
+string hobject_t::to_str() const
```
Comment (Matan-B) [inline src/test/common/test_hobject.cc:164]: Why these functions needs to be copied here and can't be re-used from `common/hobject.cc`?
Same for append_escaped, append_out_escaped

Comment (Matan-B) [reply src/test/common/test_hobject.cc:164]: Ah, this is the previous version of this code.

Comment (ronen-fr) [reply src/test/common/test_hobject.cc:164]: Interesting addition to the tests, but not a replacement. But:
- I did not find such a ctor. Maybe I missed it? and
- there are (at least) two different ways to 'stringify' an hobject: to_str() and the formatter use
 different escaping mechanism (which is interesting by itself). 

I do want to keep a fixed base line, so there is a point in copying the code in.

**Review by Matan-B — APPROVED** [review]

**Comment by ronen-fr** [comment]
Merging based on my Teuthology tests. To note:
[rfriedma-2024-02-21_19:28:31-rados:thrash-wip-rf-hobject-distro-default-smithi](https://pulpito.ceph.com/rfriedma-2024-02-21_19:28:31-rados:thrash-wip-rf-hobject-distro-default-smithi/)/7569753 failure is interesting, but not related to the PR

**Pull request merged**